### PR TITLE
Update partial-classes-and-methods.md

### LINKED ISSUE #26542: Partial Classes and Methods programming guide contradicts C# spec about generic constraints
On the "[Partial Classes and Methods - C# Programming Guide](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/partial-classes-and-methods)" page, it is said that:

> Partial methods can be generic. Constraints are put on the defining partial method declaration, and may optionally be repeated on the implementing one.

This is contradictory to what the [C# language specification](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes#partial-methods) says (and also to the actual behaviour of `dotnet`), which is:

> Corresponding type parameters in the declarations must have the same constraints (modulo differences in type parameter names).

Generic constraints ***must*** be repeated on the implementing declaration, not "optionally", as this code does not compile:

```
public partial class C {
    public partial void M<T>() where T: class;
}

public partial class C {
    public partial void M<T>() {
        
    }
}
```

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 450a248b-ab3f-ac3f-d3a5-b1e74b3355ba
* Version Independent ID: 5e05ac12-7d29-d144-a49a-e54765343c20
* Content: [Partial Classes and Methods - C# Programming Guide](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/partial-classes-and-methods)
* Content Source: [docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md](https://github.com/dotnet/docs/blob/main/docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md)
* Product: **dotnet-csharp**
* Technology: **csharp-fundamentals**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

---
[Associated WorkItem - 229457](https://dev.azure.com/msft-skilling/Content/_workitems/edit/229457)

### LINKED ISSUE #30017: Partial methods and visibility is unclear
In this article and its [companion topic](https://docs.microsoft.com/dotnet/csharp/language-reference/keywords/partial-method), it appears as if the default for a partial method with no visibility modifier is `private`.

It is also suggested that by adding a visibility modifier like `public` to the partial declaration requires that an implementation is provided.

It's also suggested that the partial declaration and the implementation signatures must match.

This leads me to believe you can't have a declaration like:

```cs
partial void OnSomethingHappened();
```

and make the implementation protected like:
```cs
protected partial void OnSomethingHappened() { ... }
```

Clarifying how this works/doesn't work with an example or more prose in the bullets would be helpful.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 450a248b-ab3f-ac3f-d3a5-b1e74b3355ba
* Version Independent ID: 5e05ac12-7d29-d144-a49a-e54765343c20
* Content: [Partial Classes and Methods - C# Programming Guide](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/partial-classes-and-methods)
* Content Source: [docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md](https://github.com/dotnet/docs/blob/main/docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md)
* Product: **dotnet-csharp**
* Technology: **csharp-fundamentals**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

---
[Associated WorkItem - 229456](https://dev.azure.com/msft-skilling/Content/_workitems/edit/229456)

### LINKED ISSUE #31318: Links are to C#6 specifications
The links in section https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/partial-classes-and-methods#c-language-specification go to a c#6 document which does not mention partial methods.

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 450a248b-ab3f-ac3f-d3a5-b1e74b3355ba
* Version Independent ID: 5e05ac12-7d29-d144-a49a-e54765343c20
* Content: [Partial Classes and Methods - C# Programming Guide](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/partial-classes-and-methods)
* Content Source: [docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md](https://github.com/dotnet/docs/blob/main/docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md)
* Product: **dotnet-csharp**
* Technology: **csharp-fundamentals**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

---
[Associated WorkItem - 229455](https://dev.azure.com/msft-skilling/Content/_workitems/edit/229455)

### PR DESCRIPTION
**Pull request by BillWagner**
Contributes to #39819

- Fixes #26542:  Fix errors in describing partial generic methods and classes.
- Fixes #30017:  Reword rules on partial methods that don't need an implementation.
- Fixes #31318:  Add additional links to the C# 7 spec for partial methods, and the C# 9 enhancements to partial methods.

Also, do a general edit pass.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md](https://github.com/dotnet/docs/blob/b2b0060799e15b7627934c182feff974719afdc0/docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md) | [Partial Classes and Methods (C# Programming Guide)](https://review.learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/partial-classes-and-methods?branch=pr-en-us-39911) |

### CODE REVIEW

File: docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md [hunk 1]
```diff
@@ -1,28 +1,27 @@
 ---
 title: "Partial Classes and Methods - C# Programming Guide"
 description: Partial classes and methods in C# split the definition of a class, a struct, an interface, or a method over two or more source files.
-ms.date: 06/21/2021
+ms.date: 03/12/2024
 helpviewer_keywords:
   - "partial methods [C#]"
   - "partial classes [C#]"
   - "C# language, partial classes and methods"
-ms.assetid: 804cecb7-62db-4f97-a99f-60975bd59fa1
```
Comment (gewarren) [inline docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md:9]: <3

File: docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md [hunk 2]
```diff
@@ -1,28 +1,27 @@
...
 ---
 # Partial Classes and Methods (C# Programming Guide)
 
-It is possible to split the definition of a [class](../../language-reference/keywords/class.md), a [struct](../../language-reference/builtin-types/struct.md), an [interface](../../language-reference/keywords/interface.md) or a method over two or more source files. Each source file contains a section of the type or method definition, and all parts are combined when the application is compiled.
+It's possible to split the definition of a [class](../../language-reference/keywords/class.md), a [struct](../../language-reference/builtin-types/struct.md), an [interface](../../language-reference/keywords/interface.md) or a method over two or more source files. Each source file contains a section of the type or method definition, and all parts are combined when the application is compiled.
```
Comment (gewarren) [inline docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md:12]:
```suggestion
It's possible to split the definition of a [class](../../language-reference/keywords/class.md), a [struct](../../language-reference/builtin-types/struct.md), an [interface](../../language-reference/keywords/interface.md), or a method over two or more source files. Each source file contains a section of the type or method definition, and all parts are combined when the application is compiled.
```

File: docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md [hunk 3]
```diff
@@ -1,28 +1,27 @@
...
 
 ## Partial Classes
 
 There are several situations when splitting a class definition is desirable:
 
-- When working on large projects, spreading a class over separate files enables multiple programmers to work on it at the same time.
-- When working with automatically generated source, code can be added to the class without having to recreate the source file. Visual Studio uses this approach when it creates Windows Forms, Web service wrapper code, and so on. You can create code that uses these classes without having to modify the file created by Visual Studio.
-- When using [source generators](../../roslyn-sdk/source-generators-overview.md) to generate additional functionality in a class.
+- Declaring a class over separate files enables multiple programmers to work on it at the same time.
+- Code can be added to the class without having to recreate the source file that includes automatically generated source. Visual Studio uses this approach when it creates Windows Forms, Web service wrapper code, and so on. You can create code that uses these classes without having to modify the file created by Visual Studio.
```
Comment (gewarren) [inline docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md:19]:
```suggestion
- You can add code to the class without having to recreate the source file that includes automatically generated source. Visual Studio uses this approach when it creates Windows Forms, Web service wrapper code, and so on. You can create code that uses these classes without having to modify the file created by Visual Studio.
```

File: docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md [hunk 4]
```diff
@@ -55,24 +54,24 @@ The following are merged from all the partial-type definitions:
...
 
 ### Restrictions
 
-There are several rules to follow when you are working with partial class definitions:
+There are several rules to follow when you're working with partial class definitions:
 
 - All partial-type definitions meant to be parts of the same type must be modified with `partial`. For example, the following class declarations generate an error:
-  [!code-csharp[AllDefinitionsMustBePartials#7](snippets/partial-classes-and-methods/Program.cs#7)]
+  :::code language="csharp" source="snippets/partial-classes-and-methods/Program.cs" id="Snippet7":::
 - The `partial` modifier can only appear immediately before the keywords `class`, `struct`, or `interface`.
```
Comment (gewarren) [inline docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md:69]:
```suggestion
- The `partial` modifier can only appear immediately before the keyword `class`, `struct`, or `interface`.
```

File: docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md [hunk 5]
```diff
@@ -89,27 +88,36 @@ For more information, see [Constraints on Type Parameters](../generics/constrain
...
-- It doesn't have any of the following modifiers [virtual](../../language-reference/keywords/virtual.md), [override](../../language-reference/keywords/override.md), [sealed](../../language-reference/keywords/sealed.md), [new](../../language-reference/keywords/new-modifier.md), or [extern](../../language-reference/keywords/extern.md).
+An implementation isn't required for a partial method when the signature obeys the following rules:
 
-Any method that doesn't conform to all those restrictions (for example, `public virtual partial void` method), must provide an implementation. That implementation may be supplied by a *source generator*.
+- The declaration doesn't include any access modifiers. The method has [`private`](../../language-reference/keywords/private.md) access by default.
+- The return type is [`void`](../../language-reference/builtin-types/void.md).
+- None of the parameters have the [`out`](../../language-reference/keywords/method-parameters.md#out-parameter-modifier) modifier.
+- The method declaration can't include any of the following modifiers:
+  - [virtual](../../language-reference/keywords/virtual.md),
```
Comment (gewarren) [inline docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md:107]: I don't think these bullets should have commas or periods.

**Review by gewarren — APPROVED** [review]

File: docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md [hunk 6]
```diff
@@ -89,27 +88,36 @@ For more information, see [Constraints on Type Parameters](../generics/constrain
...
+  - [override](../../language-reference/keywords/override.md),
+  - [sealed](../../language-reference/keywords/sealed.md),
+  - [new](../../language-reference/keywords/new-modifier.md),
+  - [extern](../../language-reference/keywords/extern.md).
```
Comment (BillWagner) [inline docs/csharp/programming-guide/classes-and-structs/partial-classes-and-methods.md:111]:
```suggestion
  - [virtual](../../language-reference/keywords/virtual.md)
  - [override](../../language-reference/keywords/override.md)
  - [sealed](../../language-reference/keywords/sealed.md)
  - [new](../../language-reference/keywords/new-modifier.md)
  - [extern](../../language-reference/keywords/extern.md)
```

**Pull request merged**